---
title: "breakpoints"
output:
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: inline
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 


```{r setup}
Sys.Date()


library(ggplot2)
library(viridis)
library(tidyr)
library(dplyr)
library(pals)
library(ggpubr)
library(tidyverse)

cmd_args <- commandArgs(trailingOnly = TRUE)

#dir organisation: intermediate, plots
```


```{r data}
#read busco full.table concatenated for all species, but use the old Menc messenoides and Mel marsaeus assembly to match the genome scans!!! 
#prep with rename_busco_output.R
#this part is the same as in rearrangement stats

#table_file <- cmd_args[1]
table_file <- "../output/busco_file.table"

busco_file <- read.table(table_file)


busco_file <- 
busco_file %>%
filter(!is.na(anc_element))

#get the right order of the taxa
bin_id_order <- c("Melitaea cinxia", "Danaus plexippus", "Mec. macrinus", "Mec. mazaeus", "Mec. menapis", "Mec. messenoides", "Mec. polymnia", "Mel. ludovica", "Mel. isocomma", "Mel. mothone", "Mel. menophilus", "Mel. marsaeus")

taxa_names_id <- data.frame(cbind(species_name=c("Melitaea cinxia", "Danaus plexippus", "Melinaea ludovica", "Melinaea isocomma", "Melinaea mothone", "Melinaea menophilus", "Melinaea marsaeus", "Mechanitis macrinus", "Mechanitis mazaeus", "Mechanitis menapis", "Mechanitis messenoides", "Mechanitis polymnia"), taxa=bin_id_order))

# taxa_names <- bin_id_order
# names(taxa_names) <- taxa_names_id$species_name

busco_file <- busco_file[order(match(busco_file$species_name, c("Melitaea cinxia", "Danaus plexippus", "Mec. macrinus", "Mec. mazaeus", "Mec. menapis", "Mec. messenoides", "Mec. polymnia", "Mel. ludovica", "Mel. isocomma", "Mel. mothone", "Mel. menophilus", "Mel. marsaeus"))),]
busco_file$species_name <- as.factor(busco_file$species_name)
busco_file$species_name <- factor(busco_file$species_name, levels = c(bin_id_order))


#remove W
busco_file <- busco_file[busco_file$chr!="W" & busco_file$chr!="W2"& busco_file$chr!="W1" & busco_file$chr!="W3",] 


```




```{r gen_breakpoint_regions_mec}

genus <- "Mec"
ref_coord <- "ilMecMess1"

#filter taxa and keep only markers present in all selected taxa
filter.df <- 
busco_file[stringr::str_detect(busco_file$taxa, genus), ] 


filter.df$taxa <- as.character(filter.df$taxa) 

#keep only markers present in all selected taxa
filter.df <- subset(filter.df, marker %in% Reduce(intersect, split(filter.df$marker, filter.df$taxa)))

#remove reference from queries
queries <- unique(filter.df[filter.df$taxa != ref_coord, "taxa"])

#make a list af all comparisons
list_comp <- data.frame(cbind(ref=c(rep(ref_coord,4),
                                    rep(queries[1],3), 
                                    rep(queries[2],2), 
                                    rep(queries[3],1)), 
                   query=c(queries, queries[2:4], queries[3:4], queries[4])))

#colours for plotting
synt_col <- rev(as.vector(c(pals::kelly(),pals::polychrome()))[2:c(1+length(unique(busco_file$anc_element)))])
#change grey
synt_col[9] <-"#85660D"
synt_col[10] <- "#66B0FF"  
synt_col[11] <- "#f4ed00"  

#this loop prints a table of breakpoints for each comparison
#plots the chromosome of the pairwise reference painted with query chromosomes and the breakpoints
#plots the genus reference with breakpoints mapped to the genus reference (ref_coord)
#appends to table of the breakpoint start and end mapped to the genus reference (ref_coord)

for (i in 1:length(list_comp$ref)) {

  ref_taxa=list_comp[i, "ref"]
  query1=list_comp[i, "query"]

ref_query1 <- 
  filter.df[order(filter.df$taxa, filter.df$chr, filter.df$start),] %>%
  filter(taxa==ref_taxa) %>%
  left_join(filter.df[filter.df$taxa==query1,], by="marker", suffix = c("", "_query1")) 


ref_query1$ref_chr_query_chr <- paste(ref_query1$chr, ref_query1$chr_query1, sep = "_")

#count the length of blocks, get a number at the end of ech block
ref_query1$a <- 0
ref_query1 <- ref_query1[order(ref_query1$taxa, ref_query1$chr, ref_query1$start),]
ref_query1$a[cumsum(rle(ref_query1$ref_chr_query_chr)$lengths)] <- rle(ref_query1$ref_chr_query_chr)$lengths

#filter groups with only one gene
#check
ref_query1 %>%
  filter(a==1)

#filter
ref_query1_filt <-  
ref_query1 %>%
  filter(!a==1)

#redo breakpoints
ref_query1_filt$a <- 0
ref_query1_filt <- ref_query1_filt[order(ref_query1_filt$taxa, ref_query1_filt$chr, ref_query1_filt$start),]
ref_query1_filt$a[cumsum(rle(ref_query1_filt$ref_chr_query_chr)$lengths)] <- rle(ref_query1_filt$ref_chr_query_chr)$lengths

#add start of block and keep only start and end of blocks
breakpoints_q <- 
ref_query1_filt %>%
  mutate(a=case_when(row_number()==1 ~ 1,
                     lag(a)>1 ~ 1,
                     .default = a)) %>%
  filter(a>0)



#filter ends
list_chr_end <-  
filter.df %>%
  group_by(chr, taxa) %>%
filter(c(start==min(start) | end==max(end)))

#ends in common ref & q1
chr_end1 <- 
list_chr_end %>% 
    ungroup() %>%
  filter(taxa %in% c(ref_taxa, query1)) %>%
  filter(marker %in% intersect(list_chr_end[list_chr_end$taxa==ref_taxa, "marker"], list_chr_end[list_chr_end$taxa==query1, "marker"])$marker) %>%
  select(marker) %>%
  unique()


chr_end1$comp <- paste(ref_taxa, query1, sep = "_")

breakpoints_q <- breakpoints_q %>%
filter(!marker %in% chr_end1$marker)

breakpoints_q$br_origin <- paste(ref_taxa, query1, sep = "_")

#check breakpoints in plot, synteny?
ref_query1 %>%
  filter(taxa==ref_taxa) %>%
  ggplot() +
  geom_point(aes(chr, start, colour=chr_query1), shape=15) +
  geom_point(data=breakpoints_q, aes(chr, start), colour="black", position = position_dodge(width = 0.5)) +
facet_wrap(~taxa, scales = "free_y") +
scale_colour_manual(name = query1,
                    values = synt_col) +
coord_flip() +
theme_pubr() +
theme(legend.position = "right",
axis.title = element_blank())

ggsave(file = paste("../output/comparisons/", genus, "/br_", ref_taxa,"_", query1, Sys.Date(), genus, ".png", sep = ""),
       height = 6,
       width = 8)

#add to common table
if (i == 1){
  breakpoints <- breakpoints_q
} else {
  breakpoints <- rbind(breakpoints, breakpoints_q)
}

write.table(breakpoints_q, file = paste("../output/comparisons/", genus, "/br_", ref_taxa,"_", query1, Sys.Date(), genus, ".table", sep = ""))

#use ref coordinates and append to concat file
br_ref <- 
left_join(breakpoints_q[,c("marker", "br_origin")], filter.df[filter.df$taxa==ref_coord,])

br_ref <- br_ref %>%
  arrange(chr, start)

br_ref <- 
br_ref %>%
  mutate(br_start=end, br_end=lead(start)) %>%
  group_by(chr) %>%
  mutate(br_start = case_when(start<1000000 ~ 1,
                              start>chr_len-1000000 ~ end,
                              start > 1000000 & start<chr_len-1000000 & start > lag(start)+1000000 ~ start,
                              .default = br_start),
         br_end = case_when(start<1000000 ~ start,
                            start>chr_len-1000000 ~ chr_len,
                              start > 1000000 & start<chr_len-1000000 & start > lag(start)+1000000 ~ end,
                            .default = br_end)) %>%
  mutate(br_length=abs(br_end-br_start),
         gene_length=abs(end-start)) %>% 
  mutate(br_length_corr = case_when(abs(br_length) > 1000000 ~ NA,
                                    .default = br_length)) %>%
  filter(!is.na(br_length_corr))



#plot breaks for qc
filter.df[order(filter.df$taxa, filter.df$chr, filter.df$start),] %>%
  filter(taxa==ref_coord) %>%
  left_join(filter.df[filter.df$taxa==query1,], by="marker", suffix = c("", "_query1")) %>%
  filter(taxa==ref_coord) %>%
  ggplot() +
  geom_point(aes(chr, start, colour=chr_query1), shape=15) +
  geom_point(data=br_ref, aes(chr, start), colour="black", position = position_dodge(width = 0.5)) +
facet_wrap(~taxa, scales = "free_y") +
scale_colour_manual(name = query1,
                    values = synt_col) +
coord_flip() +
theme_pubr() +
labs(title = paste("Break points ", ref_taxa, query1, sep = "_")) +
theme(legend.position = "right",
axis.title = element_blank())

ggsave(file = paste("../output/comparisons/", genus, "/br_", ref_taxa,"_", query1, ref_coord, Sys.Date(), genus, ".png", sep = ""),
       height = 6,
       width = 8)

  if (i==1) {
    write.table(br_ref, file = paste("../output/comparisons/", genus, "/br_all_", Sys.Date(), genus, ref_coord, ".table", sep = ""), row.names = F)
    } else {
      write.table(br_ref, file = paste("../output/comparisons/", genus, "/br_all_", Sys.Date(), genus, ref_coord, ".table", sep = ""), row.names = F, append = T, col.names = F)
  }

#shared ends
chr_end1_position <- 
left_join(chr_end1, filter.df[filter.df$taxa==ref_coord,]) %>%
  group_by(chr) %>%
  mutate(end_start = case_when(start<1000000 ~ 1,
                              start>(chr_len - 1000000) ~ start,
                              .default = start),
         end_end = case_when(start<1000000 ~ end,
                             start>(chr_len - 1000000) ~ chr_len,
                            .default = end))



if (i==1) {
    write.table(chr_end1_position, file = paste("../output/comparisons/", genus, "/shared_ends_", Sys.Date(), genus, ".table", sep = ""), row.names = F)
    } else {
      write.table(chr_end1_position, file = paste("../output/comparisons/", genus, "/shared_ends_", Sys.Date(), genus, ".table", sep = ""), row.names = F, append = T, col.names = F)
}

}
#manually check bed breakpoint regions against plots

```


```{r breakpoint_regions_mel}

genus <- "Mel"
ref_coord <- "ilMelMars1"

#filter taxa
filter.df <- 
busco_file[stringr::str_detect(busco_file$taxa, genus), ]

filter.df$taxa <- as.character(filter.df$taxa) 

#remove Mel cinxia!!
filter.df <- 
  filter.df %>% filter(!taxa=="ilMelCinx1")

#keep only markers present in all selected taxa
filter.df <- subset(filter.df, marker %in% Reduce(intersect, split(filter.df$marker, filter.df$taxa)))

#remove reference from queries
queries <- unique(filter.df[filter.df$taxa != ref_coord, "taxa"])

#make a list of comparisons
list_comp <- data.frame(cbind(ref=c(rep(ref_coord,4),rep(queries[1],3), rep(queries[2],2), rep(queries[3],1)), 
                   query=c(queries, queries[2:4], queries[3:4], queries[4])))

#this loop prints a table of breakpoints between each comparison
#plots the chromosome of the pairwise reference painted with query chromosomes and the breakpoints
#plots the genus reference with breakpoints mapped to the genus reference
#appends to table of the breakpoint start and end mapped to the genus reference

#colours for plotting
synt_col <- rev(as.vector(c(pals::kelly(),pals::polychrome()))[2:c(1+length(unique(busco_file$anc_element)))])
#change grey
synt_col[9] <-"#85660D"
synt_col[10] <- "#66B0FF"  
synt_col[11] <- "#f4ed00"  

#this loop prints a table of breakpoints for each comparison
#plots the chromosome of the pairwise reference painted with query chromosomes and the breakpoints
#plots the genus reference with breakpoints mapped to the genus reference (ref_coord)
#appends to table of the breakpoint start and end mapped to the genus reference (ref_coord)

for (i in 1:length(list_comp$ref)) {

  ref_taxa=list_comp[i, "ref"]
  query1=list_comp[i, "query"]

ref_query1 <- 
  filter.df[order(filter.df$taxa, filter.df$chr, filter.df$start),] %>%
  filter(taxa==ref_taxa) %>%
  left_join(filter.df[filter.df$taxa==query1,], by="marker", suffix = c("", "_query1")) 


ref_query1$ref_chr_query_chr <- paste(ref_query1$chr, ref_query1$chr_query1, sep = "_")

#count the length of blocks, get a number at the end of ech block
ref_query1$a <- 0
ref_query1 <- ref_query1[order(ref_query1$taxa, ref_query1$chr, ref_query1$start),]
ref_query1$a[cumsum(rle(ref_query1$ref_chr_query_chr)$lengths)] <- rle(ref_query1$ref_chr_query_chr)$lengths

#filter groups with only one gene
#check
ref_query1 %>%
  filter(a==1)

#filter
ref_query1_filt <-  
ref_query1 %>%
  filter(!a==1)

#redo breakpoints
ref_query1_filt$a <- 0
ref_query1_filt <- ref_query1_filt[order(ref_query1_filt$taxa, ref_query1_filt$chr, ref_query1_filt$start),]
ref_query1_filt$a[cumsum(rle(ref_query1_filt$ref_chr_query_chr)$lengths)] <- rle(ref_query1_filt$ref_chr_query_chr)$lengths

#add start of block and keep only start and end of blocks
breakpoints_q <- 
ref_query1_filt %>%
  mutate(a=case_when(row_number()==1 ~ 1,
                     lag(a)>1 ~ 1,
                     .default = a)) %>%
  filter(a>0)



#filter ends
list_chr_end <-  
filter.df %>%
  group_by(chr, taxa) %>%
filter(c(start==min(start) | end==max(end)))

#ends in common ref & q1
chr_end1 <- 
list_chr_end %>% 
    ungroup() %>%
  filter(taxa %in% c(ref_taxa, query1)) %>%
  filter(marker %in% intersect(list_chr_end[list_chr_end$taxa==ref_taxa, "marker"], list_chr_end[list_chr_end$taxa==query1, "marker"])$marker) %>%
  select(marker) %>%
  unique()


chr_end1$comp <- paste(ref_taxa, query1, sep = "_")

breakpoints_q <- breakpoints_q %>%
filter(!marker %in% chr_end1$marker)

breakpoints_q$br_origin <- paste(ref_taxa, query1, sep = "_")

#check breakpoints in plot, synteny?
ref_query1 %>%
  filter(taxa==ref_taxa) %>%
  ggplot() +
  geom_point(aes(chr, start, colour=chr_query1), shape=15) +
  geom_point(data=breakpoints_q, aes(chr, start), colour="black", position = position_dodge(width = 0.5)) +
facet_wrap(~taxa, scales = "free_y") +
scale_colour_manual(name = query1,
                    values = synt_col) +
coord_flip() +
theme_pubr() +
theme(legend.position = "right",
axis.title = element_blank())

ggsave(file = paste("../output/comparisons/", genus, "/br_", ref_taxa,"_", query1, Sys.Date(), genus, ".png", sep = ""),
       height = 6,
       width = 8)

#add to common table
if (i == 1){
  breakpoints <- breakpoints_q
} else {
  breakpoints <- rbind(breakpoints, breakpoints_q)
}


write.table(breakpoints_q, file = paste("../output/comparisons/", genus, "/br_", ref_taxa,"_", query1, Sys.Date(), genus, ".table", sep = ""))

#use ref coordinates and append to concat file
br_ref <- 
left_join(breakpoints_q[,c("marker", "br_origin")], filter.df[filter.df$taxa==ref_coord,])

br_ref <- br_ref %>%
  arrange(chr, start)

br_ref <- 
br_ref %>%
  mutate(br_start=end, br_end=lead(start)) %>%
  group_by(chr) %>%
  mutate(br_start = case_when(start<1000000 ~ 1,
                              start>chr_len-1000000 ~ end,
                              start > 1000000 & start<chr_len-1000000 & start > lag(start)+1000000 ~ start,
                              .default = br_start),
         br_end = case_when(start<1000000 ~ start,
                            start>chr_len-1000000 ~ chr_len,
                              start > 1000000 & start<chr_len-1000000 & start > lag(start)+1000000 ~ end,
                            .default = br_end)) %>%
  mutate(br_length=abs(br_end-br_start),
         gene_length=abs(end-start)) %>% 
  mutate(br_length_corr = case_when(abs(br_length) > 1000000 ~ NA,
                                    .default = br_length)) %>%
  filter(!is.na(br_length_corr))



#plot breaks for qc
filter.df[order(filter.df$taxa, filter.df$chr, filter.df$start),] %>%
  filter(taxa==ref_coord) %>%
  left_join(filter.df[filter.df$taxa==query1,], by="marker", suffix = c("", "_query1")) %>%
  filter(taxa==ref_coord) %>%
  ggplot() +
  geom_point(aes(chr, start, colour=chr_query1), shape=15) +
  geom_point(data=br_ref, aes(chr, start), colour="black", position = position_dodge(width = 0.5)) +
facet_wrap(~taxa, scales = "free_y") +
scale_colour_manual(name = query1,
                    values = synt_col) +
coord_flip() +
theme_pubr() +
labs(title = paste("Break points ", ref_taxa, query1, sep = "_")) +
theme(legend.position = "right",
axis.title = element_blank())

ggsave(file = paste("../output/comparisons/", genus, "/br_", ref_taxa,"_", query1, ref_coord, Sys.Date(), genus, ".png", sep = ""),
       height = 6,
       width = 8)

  if (i==1) {
    write.table(br_ref, file = paste("../output/comparisons/", genus, "/br_all_", Sys.Date(), genus, ref_coord, ".table", sep = ""), row.names = F)
    } else {
      write.table(br_ref, file = paste("../output/comparisons/", genus, "/br_all_", Sys.Date(), genus, ref_coord, ".table", sep = ""), row.names = F, append = T, col.names = F)
  }

#shared ends
chr_end1_position <- 
left_join(chr_end1, filter.df[filter.df$taxa==ref_coord,]) %>%
  group_by(chr) %>%
  mutate(end_start = case_when(start<1000000 ~ 1,
                              start>(chr_len - 1000000) ~ start,
                              .default = start),
         end_end = case_when(start<1000000 ~ end,
                             start>(chr_len - 1000000) ~ chr_len,
                            .default = end))


if (i==1) {
    write.table(chr_end1_position, file = paste("../output/comparisons/", genus, "/shared_ends_", Sys.Date(), genus, ".table", sep = ""), row.names = F)
    } else {
      write.table(chr_end1_position, file = paste("../output/comparisons/", genus, "/shared_ends_", Sys.Date(), genus, ".table", sep = ""), row.names = F, append = T, col.names = F)
}

}
#manually check bed breakpoint regions against plots


```


```{r br_points_matrix}

#number of breaks per comparison, filter a>1 removes the start position of each break point region
#note that the breakpoints mapped to the reference genomes differ due to change in position of the markers in the ref genome vs to the compared genomes that are used here

#read in breakpoint from each comparison not the one mapped to reference
genus="Mec"
path_file=paste("../output/comparisons/", genus,"/", sep="")

breakpoints_mec <-
  list.files(path = path_file, pattern = "*03Mec.table", full.names = T)[1:10] %>% 
  map_df(~read.table(., header=T)%>% mutate(chr = as.character(chr), chr_query1=as.character(chr_query1),
                                            anc_element = as.character(anc_element), anc_element_query1 =  as.character(anc_element_query1))) 

genus="Mel"
path_file=paste("../output/comparisons/", genus,"/", sep="")

breakpoints_mel <-
  list.files(path = path_file, pattern = "*03Mel.table", full.names = T)[1:10] %>% 
  map_df(~read.table(., header=T)%>% mutate(chr = as.character(chr), chr_query1=as.character(chr_query1),
                                            anc_element = as.character(anc_element), anc_element_query1 =  as.character(anc_element_query1)))


br_df <- 
rbind(breakpoints_mel,breakpoints_mec) %>%
  filter(a>1) %>%
  group_by(br_origin) %>%
  summarize(species_name, species_name_query1, n=n()) %>%
  unique()


br_df$species_name <- sub("Melinaea", "Mel.", br_df$species_name)
br_df$species_name <- sub("Mechanitis", "Mec.", br_df$species_name)

br_df$species_name <- as.factor(br_df$species_name)
br_df$species_name <- factor(br_df$species_name, levels = c("Mec. messenoides", "Mec. macrinus", "Mec. mazaeus", "Mec. menapis", "Mel. marsaeus", "Mel. ludovica","Mel. isocomma",  "Mel. mothone"))

br_df$species_name_query1 <- sub("Melinaea", "Mel.", br_df$species_name_query1)
br_df$species_name_query1 <- sub("Mechanitis", "Mec.", br_df$species_name_query1)

br_df$species_name_query1 <- as.factor(br_df$species_name_query1)
br_df$species_name_query1 <- factor(br_df$species_name_query1, levels = c("Mec. macrinus", "Mec. mazaeus", "Mec. menapis", "Mec. polymnia", "Mel. ludovica", "Mel. isocomma", "Mel. mothone", "Mel. menophilus"))

write.table(br_df, paste("../output/breakpoints_", Sys.Date(), ".table", sep = ""))
br_df <- read.table("../output/breakpoints_2024-11-06.table")
  

text_size=16
  
br_df %>%
  ungroup() %>%
  select(species_name, species_name_query1,n ) %>%
ggplot(aes(species_name, species_name_query1, fill=n)) +
  geom_tile() +
  geom_text(aes(label=n), size = text_size/2.8, fontface = "bold") +
  scale_y_discrete(limits=rev) +
  scale_x_discrete(position = "bottom") +
    scale_fill_gradient(name = "Count",
                        limits = c(min(br_df$n, na.rm = T),max(br_df$n, na.rm = T)),
                        #low = "lightblue",
                        #high = "royalblue2",
                        # low = "lightskyblue1",
                        # high = "dodgerblue3"
                        low = "#c3e6fc",
                        high = "#5E83A2"
                        ) +
    theme(axis.title = element_blank(),
          axis.text.x = element_text(size= text_size, angle = 30, face = "italic", hjust = 1, vjust = 1, colour = "grey10"),
          axis.text.y = element_text(size= text_size, face = "italic", colour = "grey10", hjust = 0),
          panel.background = element_blank(),
          panel.grid = element_line(linewidth = 0.2, colour = "grey"),
          legend.position = "inside",
          legend.direction = "horizontal",
          legend.position.inside = c(0.7,0.8),
          legend.text = element_text(size = 14),
          legend.title = element_text(size = 14)
          )
    


ggsave(
       filename  = "../output/plot_nr_rearr_blue_tex_down_legtop.pdf",
       height = 4,
       width = 5.5,
       bg = "transparent")

ggsave(
       filename  = "../output/plot_nr_rearr.png",
       height = 6,
       width = 10)

```

